codechecker_lib/database_handler.py

```diff
@@ -4,74 +4,379 @@
+                command.upgrade(cfg, "head")
+
+                session.commit()
+            else:
+                CC_META.create_all(engine)
+
+            LOG.debug('Creating new database schema done')
+            return True
+
+        except sqlalchemy.exc.SQLAlchemyError as alch_err:
+            LOG.error(str(alch_err))
+            sys.exit(1)
+
+
+    @abstractmethod
+    def start(self, wait_for_start=True, init=False):
+        '''
+        Starts the database server and initializes the database server.
+
+        On wait_for_start == True, this method returns when the server is up
+        and ready for connections. Otherwise it only starts the server and
+        returns immediately.
+
+        On init == True, this it also initializes the database data and schema
+        if needed.
+
+        On error sys.exit(1) should be called.
+        '''
+        pass
+
+
+    @abstractmethod
+    def stop(self):
+        '''
+        Terminates the database server.
+
+        On error sys.exit(1) should be called.
+        '''
+        pass
+
+
+    @abstractmethod
+    def get_connection_string(self):
+        '''
+        Returns the connection string for SQLAlchemy.
+
+        DO NOT LOG THE CONNECTION STRING BECAUSE IT MAY CONTAIN THE PASSWORD
+        FOR THE DATABASE!
+        '''
+        pass
+
+
+    @classmethod
+    def from_cmdline_args(cls, args, context, env=None):
+        '''
+        Normally only this method is called form outside of this module in
+        order to instance the proper server implementation.
+
+        Parameters:
+            args: the command line arguments from CodeChecker.py
+            context: a Context object
+            env: a run environment dictionary.
+        '''
+
+        if not host_check.check_sql_driver(args.sqlite):
+            LOG.error("SQL driver error")
+            sys.exit(1)
+
+        if args.sqlite:
+            LOG.debug("Using SQLiteDatabase")
+            return SQLiteDatabase(context, run_env=env)
+        else:
+            LOG.debug("Using PostgreSQLServer")
+            return PostgreSQLServer(context,
+                                    args.dbaddress,
+                                    args.dbport,
+                                    args.dbusername,
+                                    args.dbname,
+                                    run_env=env)
+
+class PostgreSQLServer(SQLServer):
+    '''
+    Handler for PostgreSQL.
+    '''
+
+    def __init__(self, context, host, port, user, database, password = None, run_env=None):
+        super(PostgreSQLServer, self).__init__(context.migration_root)
+
+        self.path = os.path.join(context.codechecker_workspace, 'pgsql_data')
+        self.host = host
+        self.port = port
+        self.user = user
+        self.database = database
+        self.password = password
+        self.run_env = run_env
+
+        self.proc = None
+
+    def _call_command(self, command):
```

Comment:
call_command could be moved out to util module so other modules can use it
(for example get_checker_list function in the analyzer module).
If output and error messages required only in a blocking way the subprocess.check_output function can be used.